scRNA-seq/05-clustering_markers_scRNA.Rmd

```diff
@@ -0,0 +1,400 @@
+`purrr::map()` will then apply the expression to each element of the list, and return the results as a new list.
+
+```{r head_markernames, live = TRUE}
+# Get the first few row names of each table with a purrr function.
+purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr
+           ~ rownames(head(.x))) # our custom function.
+```
+
+Another variant is `purrr::imap()`, which allows us to use the names of the list elements in our function. 
+(Try `names(markers)` to see the names for the list we are working with now.)
+We will use that here to name output files where we will print each of the marker tables, one for each cell type.
+We are again defining a custom function within the call to `purrr:imap()` using the `~` syntax, but this time we can use `.x` for the list elements and `.y` for their names.
+Because we don't know the identies of the clusters we identified, these are just the cluster numbers for now.
+
+Making file names from numbers can be a a bit fraught, as we really want them to sort in numerical order, but many systems will sort by alphabetical order.
+Unfortunately, that would tend to sort 10-19 before 2, 20-29 before 3, etc. 
+To solve this, we are using the `sprintf()` function, which allows us to specify the format of a printed string.
+In this case, we are using the signifier syntax of `%02d` to tell it that we will want to insert (`%`) a number (`d`), with two digits and leading zeros.
+To see what this does a bit more simply, you might want try it out with the following command in the console: 
+`sprintf("%02d", 1:10)`
+
+In addition to writing the tables out, we are saving the data frames we created as a new list object that we can use in the next step.
+
+```{r write_tables}
+marker_df_list <- purrr::imap(
+  as.list(markers), # convert markers to a 'regular' list for purrr
+  ~ as.data.frame(.x) %>% # convert the table to a data frame
+    tibble::rownames_to_column("gene") %>% # make genes a new column
+    dplyr::arrange(FDR) %>% # sort to be sure small FDR are first
+    readr::write_tsv( # write each data frame to a file
+      file.path(marker_dir, # construct the output path
+                sprintf("cluster%02d_markers.tsv", as.integer(.y)) # format cluster numbers in file names with leading zeros
+                )
+      )
+  )
+```
+
+`purrr::map_df()` is another nice function in `purrr`. 
+It applies the operations in the second argument and combines the output into a single data frame.
+Optionally, it adds a column with the name specified by the `.id` argument to differentiate the rows that came from each element of the input list.
+
+```{r top_markers, live = TRUE}
+# Make a data frame of the top 20 markers for each cell type
+top_markers <- purrr::map_df(marker_df_list,
+                             head, 
+                             n = 20,
+                             .id = "cluster")
+```
+
+### Plotting marker gene expression
```

Comment:
If we're going to look at a single cluster, is there a way to skip `purrr::map_df()`?